### ISSUE TITLE
Solver misses sign reversal on simple example

### ISSUE THREAD

**Issue by svillemot** [thomasokken/free42#10]
I was following the example given p. 116 of the “Example and techniques” manual (solving the function `f(x)=x/(x²-6)`, with initial guesses of 2.3 and 2.7).

Apparently the original HP-42S would detect the sign reversal (the function has a pole at √6). Free42 instead indicates success (returning a variable value close to √6, with of course a very large function value).

Depending on how close you aim to be to the original HP-42S solver, this may or may not qualify as a bug.

**Comment by thomasokken** [comment]
Free42 only detects sign reversals in one very specific case, namely, when between the values that return a positive result and those that return a negative result, there is at least one that returns no result (i.e. an error, like a division by zero). It does not detect a sign reversal when all values return a result without error, even if the values diverge around the sign reversal.

I agree it would be better if Free42 would indicate a sign reversal in the given case, but I haven't yet found a way to distinguish between such sign reversals and legitimate zeros (where the function just happens to never evaluate to zero exactly) in a reliable manner. If I knew a way to make that distinction, I would be happy to modify the algorithm.

**Comment by jackokring** [comment]
The sign reversal can be detected by noting the change in root in f(x)+C and f(x)-C, whereas a pole will infinity load the value of x to have massive inertia to stay put. Perhaps this helps. Also of note is the convergence behaviour about the root for the sum of the product of the function at the bisection ends.

**Comment by thomasokken** [comment]
I'm not following. Could you illustrate with some pseudocode?

**Comment by jackokring** [comment]
Suppose the solver solves y=f(x) and y=1/f(x) returning the same. a comparative run of f(x-d)*f(x+d) where d approaches zero would show a either a zero or pole value. As d approaches zero in an infinity of time approximated by a convergence of the series, the root will converge to zero and the pole will converge to minus infinity.and https://en.wikipedia.org/wiki/Series_acceleration improves the result without going all the way to adding an infinite series.

That might be a better way as dual roots turning complex with offsets of a constant f(x)+C would be more difficult to handle than using the interval bounds to like a variable d as f(upper)*f(lower) would have many properties of f(x-d)*f(x+d) in the limit as d goes to zero.

**Comment by jackokring** [comment]
def solve(f)
 repeat
  x(estimateCount = n, estimateSide = m) = inbuiltSolve(f)
 until size(m) > 1 //bisection root between x(n, 0) and x(n, 1)
 q(n) = p(n) = f(x(n, 0))*f(x(n, 1))
 repeat
  x(n, m) = inbuiltSolve(f)
  p(n) += f(x(n, 0))*f(x(n, 1))
  z(n, m) = inbuiltSolve(1/f)
  q(n) += 1/f(z(n, 0))*1/f(z(n, 1))
  //perhaps use extra z estimates as well as x estimates to improve convergence
 until solvedStop()
 if(q(n) < p(n)) show rootSolution(x(n)) else show poleSolution(x(n))

**Comment by thomasokken** [comment]
I was looking at the solver code the other day for Plus42-related reasons, and it occurred to me that this issue may be at least partly caused by the logic that deals with steep secants being a bit too eager to declare something a root. See core_math1.cc, line 859. I'll have to look into this.

**Comment by thomasokken** [comment]
No, that's not it. The x/(x^2-6) test case, with starting values 2.3 and 2.7, ends at case 6, line 952, i.e. an orderly termination after a Ridders refinement step.
So the challenge still looks to be to somehow determine that the secant is suspiciously steep, for unknown values of "suspicious"...

**Comment by jackokring** [comment]
Hi

So as the difference between the last and the new root esimate can be added
to the last to get the new, the first plus a series of differences make an
infinite series (converging?) On a root.

Given about 7 terms of this series a fast "series acceleration" can be
done. To estimate the infinity of terms sum.

Does this quickly indicate a root or a divergence?

**Comment by thomasokken** [comment]
I implemented a new experimental sign reversal detection algorithm. It simply tracks f(x2) - f(x1): if that gap gets larger during each of the last 3 iterations before x1 and x2 become equal, SOLVE reports Sign Reversal.
That magic number 3 is just a guess. It can't be too low or you'll get false positives, and it can't be too high or you'll never detect any sign reversals. 3 seems reasonable, and in any case, it works with svillemot's test case.

The code is here: 5277dca53ab38c86d04cbd029568efcb7fd31125

I uploaded test builds for Windows to https://thomasokken.com/free42/download/test/
